File: src/app/api/video/services/video.ts
// Video processing service using ffmpeg
import { spawn } from 'child_process';
import fs from 'fs';
import path from 'path';
import { AudioFileData } from '../types';
import { combineAudioBuffers } from './audio';

interface SubtitleConfig {
  fadeInDuration: number;
  fadeOutDuration: number;
  scaleAnimation: boolean;
  dynamicPositioning: boolean;
  maxSimultaneousLines: number;
}

interface VideoConfig {
  subtitleConfig: SubtitleConfig;
}

export async function createFinalVideoWithBuffers(
  videoPath: string,
  stewieImagePath: string,
  peterImagePath: string,
  audioData: AudioFileData[],
  subtitleContent: string,
  characterTimeline: Array<{ character: 'stewie' | 'peter'; startTime: number; endTime: number }>,
  duration: number,
  imageOverlays?: Array<{
    buffer: Buffer;
    startTime: number;
    endTime: number;
    description: string;
  }>,
  config?: VideoConfig
): Promise<Buffer> {
  const { fadeInDuration, fadeOutDuration } = config?.subtitleConfig ?? {
    fadeInDuration: 0.05,
    fadeOutDuration: 0.05,
    scaleAnimation: false,
    dynamicPositioning: true,
    maxSimultaneousLines: 1
  };

  // Write subtitle content to temp file (FFmpeg requires file path for subtitles)
  const subtitlePath = path.join('/tmp', `subtitles_${Date.now()}.ass`);
  await fs.promises.writeFile(subtitlePath, subtitleContent);

  // Combine audio buffers
  const combinedAudioBuffer = await combineAudioBuffers(audioData);
  const combinedAudioPath = path.join('/tmp', `combined_audio_${Date.now()}.wav`);
  await fs.promises.writeFile(combinedAudioPath, combinedAudioBuffer);

  // Write image overlay buffers to temp files if needed
  const overlayTempFiles: string[] = [];
  try {
    for (let i = 0; i < (imageOverlays?.length || 0); i++) {
      const overlay = imageOverlays![i];
      const tempImagePath = path.join('/tmp', `overlay_${Date.now()}_${i}.png`);
      await fs.promises.writeFile(tempImagePath, overlay.buffer);
      overlayTempFiles.push(tempImagePath);
    }

    // Create character overlay expressions
    const stewieOverlay = characterTimeline
      .filter(t => t.character === 'stewie')
      .map(t => `between(t,${t.startTime},${t.endTime})*if(between(t,${t.startTime},${t.startTime + fadeInDuration}),(t-${t.startTime})/${fadeInDuration},if(between(t,${t.endTime - fadeOutDuration},${t.endTime}),(${t.endTime}-t)/${fadeOutDuration},1))`)
      .join('+');

    const peterOverlay = characterTimeline
      .filter(t => t.character === 'peter')
      .map(t => `between(t,${t.startTime},${t.endTime})*if(between(t,${t.startTime},${t.startTime + fadeInDuration}),(t-${t.startTime})/${fadeInDuration},if(between(t,${t.endTime - fadeOutDuration},${t.endTime}),(${t.endTime}-t)/${fadeOutDuration},1))`)
      .join('+');

    // Build the FFmpeg command
    const inputs = [
      '-t', duration.toString(),
      '-i', videoPath,
      '-i', stewieImagePath,
      '-i', peterImagePath,
      '-i', combinedAudioPath
    ];

    const imageFilterChain: string[] = [];
    if (imageOverlays && imageOverlays.length > 0) {
      imageOverlays.forEach((overlay, index) => {
        inputs.push('-i', overlayTempFiles[index]);
        const inputIndex = 4 + index; // Start after audio input (index 3)
        const scaledLabel = `img_${index}_scaled`;
        
        imageFilterChain.push(`[${inputIndex}:v]scale=600:-1[${scaledLabel}]`);
      });
    }

    const filterChain = [
      // Scale character images
      '[1:v]scale=-1:700[stewie_img]',
      '[2:v]scale=-1:700[peter_img]',
      
      // Add character overlays with dynamic positioning
      `[0:v][stewie_img]overlay=400:H-h-30:enable='${stewieOverlay}'[with_stewie]`,
      `[with_stewie][peter_img]overlay=-300:H-h-30:enable='${peterOverlay}'[with_characters]`,
      
      // Add image scaling for overlays
      ...imageFilterChain,
      
      // Add image overlays
      ...(imageOverlays && imageOverlays.length > 0 ? 
        imageOverlays.map((overlay, index) => {
          const inputLabel = index === 0 ? 'with_characters' : `with_overlay_${index - 1}`;
          const outputLabel = index === imageOverlays.length - 1 ? 'with_overlays' : `with_overlay_${index}`;
          const scaledLabel = `img_${index}_scaled`;
          const overlayEnable = `between(t,${overlay.startTime},${overlay.endTime})`;
          
          return `[${inputLabel}][${scaledLabel}]overlay=(W-w)/2:100:enable='${overlayEnable}'[${outputLabel}]`;
        }) : []
      ),
      
      // Add subtitles
      `[${imageOverlays && imageOverlays.length > 0 ? 'with_overlays' : 'with_characters'}]subtitles='${subtitlePath}':force_style='FontName=Arial Black,Fontsize=140,PrimaryColour=&H00FFFFFF,BorderStyle=1,Outline=8,Shadow=3,Alignment=2'[final]`
    ];

    const outputPath = path.join('/tmp', `output_${Date.now()}.mp4`);

    const command = [
      ...inputs,
      '-y',
      '-filter_complex',
      filterChain.join(';'),
      
      // Map the final video and audio streams
      '-map', '[final]',
      '-map', '3:a',
      
      // Video encoding settings
      '-c:v', 'libx264',
      '-preset', 'slow',
      '-crf', '18',
      '-c:a', 'aac',
      '-s', '1080x1920',
      '-r', '30',
      '-shortest',
      '-y',
      outputPath
    ];

    await new Promise((resolve, reject) => {
      const ffmpeg = spawn('ffmpeg', command);

      ffmpeg.stderr.on('data', (data) => {
        if (data.toString().includes('frame=')) {
          process.stdout.write('.');
        }
      });

      ffmpeg.on('close', (code) => {
        if (code === 0) {
          console.log('\nVideo processing completed successfully');
          resolve(undefined);
        } else {
          reject(new Error(`FFmpeg process exited with code ${code}`));
        }
      });

      ffmpeg.on('error', (err) => {
        reject(err);
      });
    });

    // Read the output video as buffer
    const videoBuffer = await fs.promises.readFile(outputPath);

    // Cleanup temp files
    await Promise.all([
      fs.promises.unlink(subtitlePath).catch(() => {}),
      fs.promises.unlink(combinedAudioPath).catch(() => {}),
      fs.promises.unlink(outputPath).catch(() => {}),
      ...overlayTempFiles.map(file => fs.promises.unlink(file).catch(() => {}))
    ]);

    return videoBuffer;

  } catch (error) {
    // Cleanup on error
    await Promise.all([
      fs.promises.unlink(subtitlePath).catch(() => {}),
      fs.promises.unlink(combinedAudioPath).catch(() => {}),
      ...overlayTempFiles.map(file => fs.promises.unlink(file).catch(() => {}))
    ]);
    throw error;
  }
}

// Keep the original function for backward compatibility if needed
export async function createFinalVideo(
  videoPath: string,
  stewieImagePath: string,
  peterImagePath: string,
  audioPath: string,
  subtitlePath: string,
  characterTimeline: Array<{ character: 'stewie' | 'peter'; startTime: number; endTime: number }>,
  duration: number,
  outputPath: string,
  imageOverlays?: Array<{
    imagePath: string;
    startTime: number;
    endTime: number;
    description: string;
  }>,
  config?: VideoConfig
): Promise<void> {
  // Default configuration
  const { fadeInDuration, fadeOutDuration } = config?.subtitleConfig ?? {
    fadeInDuration: 0.05,
    fadeOutDuration: 0.05,
    scaleAnimation: false,
    dynamicPositioning: true,
    maxSimultaneousLines: 1
  };

  // Create character overlay expressions
  const stewieOverlay = characterTimeline
    .filter(t => t.character === 'stewie')
    .map(t => `between(t,${t.startTime},${t.endTime})*if(between(t,${t.startTime},${t.startTime + fadeInDuration}),(t-${t.startTime})/${fadeInDuration},if(between(t,${t.endTime - fadeOutDuration},${t.endTime}),(${t.endTime}-t)/${fadeOutDuration},1))`)
    .join('+');

  const peterOverlay = characterTimeline
    .filter(t => t.character === 'peter')
    .map(t => `between(t,${t.startTime},${t.endTime})*if(between(t,${t.startTime},${t.startTime + fadeInDuration}),(t-${t.startTime})/${fadeInDuration},if(between(t,${t.endTime - fadeOutDuration},${t.endTime}),(${t.endTime}-t)/${fadeOutDuration},1))`)
    .join('+');

  // Build the FFmpeg command with improved subtitle rendering and image overlays
  const inputs = [
    '-t', duration.toString(),
    '-i', videoPath,
    '-i', stewieImagePath,
    '-i', peterImagePath,
    '-i', audioPath
  ];

const imageFilterChain: string[] = [];
  if (imageOverlays && imageOverlays.length > 0) {
    imageOverlays.forEach((overlay, index) => {
      inputs.push('-i', overlay.imagePath);
      const inputIndex = 4 + index; // Start after audio input (index 3)
      const scaledLabel = `img_${index}_scaled`;
      
      // Scale image to be bigger and positioned above text
      imageFilterChain.push(`[${inputIndex}:v]scale=600:-1[${scaledLabel}]`);
    });
  }

  const filterChain = [
    // Scale character images
    '[1:v]scale=-1:700[stewie_img]',
    '[2:v]scale=-1:700[peter_img]',
    
    // Add character overlays with dynamic positioning
    `[0:v][stewie_img]overlay=400:H-h-30:enable='${stewieOverlay}'[with_stewie]`,
    `[with_stewie][peter_img]overlay=-300:H-h-30:enable='${peterOverlay}'[with_characters]`,
    
    // Add image scaling for overlays
    ...imageFilterChain,
    
    // Add image overlays
    ...(imageOverlays && imageOverlays.length > 0 ? 
      imageOverlays.map((overlay, index) => {
        const inputLabel = index === 0 ? 'with_characters' : `with_overlay_${index - 1}`;
        const outputLabel = index === imageOverlays.length - 1 ? 'with_overlays' : `with_overlay_${index}`;
        const scaledLabel = `img_${index}_scaled`;
        const overlayEnable = `between(t,${overlay.startTime},${overlay.endTime})`;
        
        return `[${inputLabel}][${scaledLabel}]overlay=(W-w)/2:100:enable='${overlayEnable}'[${outputLabel}]`;
      }) : []
    ),
    
    // Add subtitles with center alignment and exact positioning
    `[${imageOverlays && imageOverlays.length > 0 ? 'with_overlays' : 'with_characters'}]subtitles='${subtitlePath}':force_style='FontName=Arial Black,Fontsize=140,PrimaryColour=&H00FFFFFF,BorderStyle=1,Outline=8,Shadow=3,Alignment=2'[final]`
  ];

  const command = [
    ...inputs,
    '-y',
    '-filter_complex',
    filterChain.join(';'),
    
    // Map the final video and audio streams
    '-map', '[final]',
    '-map', '3:a',
    
    // Video encoding settings for smooth text rendering
    '-c:v', 'libx264',
    '-preset', 'slow', // Better quality for text
    '-crf', '18', // High quality
    '-c:a', 'aac',
    '-s', '1080x1920',
    '-r', '30',
    '-shortest',
    '-y',
    outputPath
  ];


  return new Promise((resolve, reject) => {
    const ffmpeg = spawn('ffmpeg', command);

    ffmpeg.stderr.on('data', (data) => {
      // Log progress but avoid excessive logging
      if (data.toString().includes('frame=')) {
        process.stdout.write('.');
      }
    });

    ffmpeg.on('close', (code) => {
      if (code === 0) {
        console.log('\nVideo processing completed successfully');
        resolve(undefined);
      } else {
        reject(new Error(`FFmpeg process exited with code ${code}`));
      }
    });

    ffmpeg.on('error', (err) => {
      reject(err);
    });
  });
} 

File: rvc/video.py
import os
import sys
import tempfile
import subprocess
from scipy.io import wavfile
import numpy as np
import time
import uuid
import pathlib
import requests
import logging
import json
from typing import List, Dict, Optional, Tuple
import asyncio
from dataclasses import dataclass

@dataclass
class AudioFileData:
    buffer: bytes
    duration: float
    character: str

@dataclass
class CharacterTimeline:
    character: str
    start_time: float
    end_time: float

@dataclass
class ImageOverlay:
    buffer: bytes
    start_time: float
    end_time: float
    description: str

@dataclass
class SubtitleConfig:
    fade_in_duration: float = 0.05
    fade_out_duration: float = 0.05
    scale_animation: bool = False
    dynamic_positioning: bool = True
    max_simultaneous_lines: int = 1

@dataclass
class VideoConfig:
    subtitle_config: SubtitleConfig = None
    
    def __post_init__(self):
        if self.subtitle_config is None:
            self.subtitle_config = SubtitleConfig()

async def combine_audio_buffers(audio_data: List[AudioFileData]) -> bytes:
    """Combine multiple audio buffers into a single WAV file"""
    if not audio_data:
        raise ValueError("No audio data provided")
    
    # Write buffers to temporary files
    temp_files = []
    try:
        for i, data in enumerate(audio_data):
            temp_path = f"/tmp/audio_part_{uuid.uuid4()}.wav"
            with open(temp_path, 'wb') as f:
                f.write(data.buffer)
            temp_files.append(temp_path)
        
        # Use ffmpeg to concatenate audio files
        output_path = f"/tmp/combined_audio_{uuid.uuid4()}.wav"
        
        # Create concat file list
        concat_list_path = f"/tmp/concat_list_{uuid.uuid4()}.txt"
        with open(concat_list_path, 'w') as f:
            for temp_file in temp_files:
                f.write(f"file '{temp_file}'\n")
        
        # Run ffmpeg concat
        cmd = [
            'ffmpeg', '-y',
            '-f', 'concat',
            '-safe', '0',
            '-i', concat_list_path,
            '-c', 'copy',
            output_path
        ]
        
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        
        stdout, stderr = await process.communicate()
        
        if process.returncode != 0:
            raise RuntimeError(f"FFmpeg concat failed: {stderr.decode()}")
        
        # Read combined audio
        with open(output_path, 'rb') as f:
            combined_buffer = f.read()
        
        # Cleanup
        os.unlink(output_path)
        os.unlink(concat_list_path)
        
        return combined_buffer
        
    finally:
        # Cleanup temp files
        for temp_file in temp_files:
            try:
                os.unlink(temp_file)
            except:
                pass

async def create_final_video_with_buffers(
    video_path: str,
    stewie_image_path: str,
    peter_image_path: str,
    audio_data: List[AudioFileData],
    subtitle_content: str,
    character_timeline: List[CharacterTimeline],
    duration: float,
    image_overlays: Optional[List[ImageOverlay]] = None,
    config: Optional[VideoConfig] = None
) -> bytes:
    """Create final video with character overlays, subtitles, and image overlays"""
    
    if config is None:
        config = VideoConfig()
    
    fade_in_duration = config.subtitle_config.fade_in_duration
    fade_out_duration = config.subtitle_config.fade_out_duration
    
    # Write subtitle content to temp file
    subtitle_path = f"/tmp/subtitles_{int(time.time())}_{uuid.uuid4()}.ass"
    with open(subtitle_path, 'w') as f:
        f.write(subtitle_content)
    
    # Combine audio buffers
    combined_audio_buffer = await combine_audio_buffers(audio_data)
    combined_audio_path = f"/tmp/combined_audio_{int(time.time())}_{uuid.uuid4()}.wav"
    with open(combined_audio_path, 'wb') as f:
        f.write(combined_audio_buffer)
    
    # Write image overlay buffers to temp files
    overlay_temp_files = []
    try:
        if image_overlays:
            for i, overlay in enumerate(image_overlays):
                temp_image_path = f"/tmp/overlay_{int(time.time())}_{i}_{uuid.uuid4()}.png"
                with open(temp_image_path, 'wb') as f:
                    f.write(overlay.buffer)
                overlay_temp_files.append(temp_image_path)
        
        # Create character overlay expressions
        stewie_segments = [t for t in character_timeline if t.character == 'stewie']
        peter_segments = [t for t in character_timeline if t.character == 'peter']
        
        stewie_overlay = '+'.join([
            f"between(t,{t.start_time},{t.end_time})*"
            f"if(between(t,{t.start_time},{t.start_time + fade_in_duration}),"
            f"(t-{t.start_time})/{fade_in_duration},"
            f"if(between(t,{t.end_time - fade_out_duration},{t.end_time}),"
            f"({t.end_time}-t)/{fade_out_duration},1))"
            for t in stewie_segments
        ]) or '0'
        
        peter_overlay = '+'.join([
            f"between(t,{t.start_time},{t.end_time})*"
            f"if(between(t,{t.start_time},{t.start_time + fade_in_duration}),"
            f"(t-{t.start_time})/{fade_in_duration},"
            f"if(between(t,{t.end_time - fade_out_duration},{t.end_time}),"
            f"({t.end_time}-t)/{fade_out_duration},1))"
            for t in peter_segments
        ]) or '0'
        
        # Build FFmpeg command
        inputs = [
            '-t', str(duration),
            '-i', video_path,
            '-i', stewie_image_path,
            '-i', peter_image_path,
            '-i', combined_audio_path
        ]
        
        image_filter_chain = []
        if image_overlays:
            for i, (overlay, temp_file) in enumerate(zip(image_overlays, overlay_temp_files)):
                inputs.extend(['-i', temp_file])
                input_index = 4 + i
                scaled_label = f"img_{i}_scaled"
                image_filter_chain.append(f"[{input_index}:v]scale=600:-1[{scaled_label}]")
        
        # Build filter chain
        filter_parts = [
            # Scale character images
            '[1:v]scale=-1:700[stewie_img]',
            '[2:v]scale=-1:700[peter_img]',
            
            # Add character overlays
            f"[0:v][stewie_img]overlay=400:H-h-30:enable='{stewie_overlay}'[with_stewie]",
            f"[with_stewie][peter_img]overlay=-300:H-h-30:enable='{peter_overlay}'[with_characters]",
        ]
        
        # Add image scaling
        filter_parts.extend(image_filter_chain)
        
        # Add image overlays
        if image_overlays:
            for i, overlay in enumerate(image_overlays):
                input_label = 'with_characters' if i == 0 else f'with_overlay_{i-1}'
                output_label = 'with_overlays' if i == len(image_overlays) - 1 else f'with_overlay_{i}'
                scaled_label = f'img_{i}_scaled'
                overlay_enable = f"between(t,{overlay.start_time},{overlay.end_time})"
                
                filter_parts.append(
                    f"[{input_label}][{scaled_label}]overlay=(W-w)/2:100:enable='{overlay_enable}'[{output_label}]"
                )
        
        # Add subtitles
        final_input = 'with_overlays' if image_overlays else 'with_characters'
        filter_parts.append(
            f"[{final_input}]subtitles='{subtitle_path}':"
            f"force_style='FontName=Arial Black,Fontsize=140,PrimaryColour=&H00FFFFFF,"
            f"BorderStyle=1,Outline=8,Shadow=3,Alignment=2,MarginV=600'[final]"
        )
        
        filter_complex = ';'.join(filter_parts)
        
        output_path = f"/tmp/output_{int(time.time())}_{uuid.uuid4()}.mp4"
        
        command = [
            'ffmpeg',
            *inputs,
            '-y',
            '-filter_complex', filter_complex,
            '-map', '[final]',
            '-map', '3:a',
            '-c:v', 'libx264',
            '-preset', 'slow',
            '-crf', '18',
            '-c:a', 'aac',
            '-s', '1080x1920',
            '-r', '30',
            '-shortest',
            '-y',
            output_path
        ]
        
        # Run FFmpeg
        process = await asyncio.create_subprocess_exec(
            *command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        
        stdout, stderr = await process.communicate()
        
        if process.returncode != 0:
            raise RuntimeError(f"FFmpeg processing failed: {stderr.decode()}")
        
        # Read output video
        with open(output_path, 'rb') as f:
            video_buffer = f.read()
        
        # Cleanup
        os.unlink(output_path)
        
        return video_buffer
        
    finally:
        # Cleanup temp files
        for temp_file in [subtitle_path, combined_audio_path] + overlay_temp_files:
            try:
                os.unlink(temp_file)
            except:
                pass